libcaf_core/caf/actor_from_state.hpp

```diff
@@ -0,0 +1,101 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#pragma once
+
+#include "caf/detail/type_traits.hpp"
+#include "caf/event_based_actor.hpp"
+#include "caf/fwd.hpp"
+#include "caf/typed_event_based_actor.hpp"
+
+#include <new>
+#include <type_traits>
+#include <utility>
+
+namespace caf::detail {
+
+/// An event-based actor with managed state. The state is constructed separately
+/// and destroyed when the actor calls `quit`.
+template <class State, class Base>
+class actor_from_state_impl : public Base {
+public:
+  using super = Base;
+
+  friend actor_from_state_t<State>;
+
+  explicit actor_from_state_impl(actor_config& cfg) : super(cfg) {
+    // nop
+  }
+
+  void on_exit() override {
+    if (has_state_)
+      state.~State();
+  }
+
+  union {
+    State state;
+  };
+
+private:
+  bool has_state_ = false;
+};
+
+template <class T>
+struct actor_from_state_impl_base;
+
+template <>
+struct actor_from_state_impl_base<behavior> {
+  using type = event_based_actor;
+};
+
+template <class... Ts>
+struct actor_from_state_impl_base<typed_behavior<Ts...>> {
+  using type = typed_event_based_actor<Ts...>;
+};
+
+} // namespace caf::detail
+
+namespace caf {
+
+/// Helper class for automating the creation of an event-based actor with
+/// managed state.
+template <class State>
+struct actor_from_state_t {
+  /// The behavior type of the actor.
+  using behavior_type = decltype(std::declval<State*>()->make_behavior());
+
+  static_assert(detail::is_behavior_v<behavior_type>,
+                "State::make_behavior() must return a behavior");
```

Comment:
I'm supposing we don't support `make_behavior` returning `void` and calling `self->become` internally for class based actors?

Reply:
That's not the use case for this utility. We need the behavior type to deduce the proper handle for the actor and this strikes me as the natural place to do so.